src/ATen/test/broadcast_test.cpp

```diff
@@ -0,0 +1,171 @@
+#include "ATen/ATen.h"
+
+using namespace at;
+
+int main() {
+  Type & T = CPU(kFloat);
+
+  // 0) pre-req tests:
+  // can't expand empty tensor
+  try {
+    auto empty = T.randn({0});
+    empty.expand({3});
+    assert(false);
+  } catch(std::runtime_error &e) {}
+
+  // can't "expand" a Tensor to  Scalar
+  try {
+    auto tScalar = T.ones({1});
+    tScalar.get()->maybeScalar(true);
+    tScalar.expand({});
+    assert(false);
+  } catch (std::runtime_error &e) {}
+
+  // 1) out-place function with 2 args
+  {
+    // basic
+    auto a = T.randn({3, 1});
+    auto b = T.randn({5});
+    std::vector<int64_t> expanded_sizes = {3, 5};
+    assert((a + b).equal(a.expand(expanded_sizes) + b.expand(expanded_sizes)));
+
+    // with scalar
+    auto aScalar = T.ones({1});
+    aScalar.get()->maybeScalar(true);
+    b = T.randn({3, 5});
+    assert((aScalar + b).equal(aScalar.expand(b.sizes()) + b.expand(b.sizes())));
```

Comment:
Can we document what "Scalar" expansion means? I think any Scalar changes will need Zach's okay, I'm not confident enough to review them.

Reply:
it just means expand a 0-dim (non-empty) tensor.